File: src/agentcap/provider.py
"""Identify the inference backend behind an upstream URL.

Hostname classification (:func:`_hostname_fallback`) +
HF Router sub-provider pin (:func:`refine_for_sub_provider`).
:func:`probe` is the richer (network) variant — issues parallel
GETs to well-known introspection endpoints, never raises.
"""

from __future__ import annotations

import concurrent.futures
import ipaddress
import time
from typing import Any
from urllib.parse import urlparse

import httpx


# Reverse proxies / custom domains won't match; the probe path catches those.
_HOSTNAME_TO_PROVIDER: dict[str, str] = {
    "router.huggingface.co": "hf-router",
    "api.openai.com": "openai",
    "api.together.xyz": "together",
    "api.anthropic.com": "anthropic",
    "api.cerebras.ai": "cerebras",
    "api.fireworks.ai": "fireworks",
    "api.groq.com": "groq",
}


def _base_root(upstream_url: str) -> str:
    # Introspection endpoints (/props, /info, ...) live under the
    # server root, not /v1.
    base = upstream_url.rstrip("/")
    if base.endswith("/v1"):
        base = base[:-3]
    return base


def _hostname_fallback(upstream_url: str) -> str:
    host = (urlparse(upstream_url).hostname or "").lower()
    if not host:
        return "unknown"
    if host in _HOSTNAME_TO_PROVIDER:
        return _HOSTNAME_TO_PROVIDER[host]
    if host in ("localhost", "::1"):
        return "local"
    try:
        ip = ipaddress.ip_address(host)
        return "local" if (ip.is_loopback or ip.is_private) else host
    except ValueError:
        pass
    parts = host.split(".")
    return parts[-2] if len(parts) >= 2 else host


def _try_get(url: str, headers: dict, timeout: float) -> dict | None:
    try:
        r = httpx.get(url, headers=headers, timeout=timeout)
    except (httpx.HTTPError, OSError):
        return None
    if r.status_code != 200:
        return None
    ct = r.headers.get("content-type", "")
    out: dict[str, Any] = {"headers": {k.lower(): v for k, v in r.headers.items()}}
    try:
        if "json" in ct:
            out["body"] = r.json()
        else:
            out["text"] = r.text[:4096]
    except Exception:
        return None
    return out


def probe(
    upstream_url: str,
    *,
    api_key: str | None = None,
    timeout: float = 3.0,
) -> dict:
    """Probe an OpenAI-compat upstream. Never raises."""
    root = _base_root(upstream_url)
    headers = {"Authorization": f"Bearer {api_key}"} if api_key else {}
    targets = {
        "props":   f"{root}/props",          # llama.cpp
        "info":    f"{root}/info",           # TGI
        "version": f"{root}/version",        # vLLM
        "models":  f"{root}/v1/models",
        "metrics": f"{root}/metrics",
    }
    endpoints: dict[str, dict] = {}
    with concurrent.futures.ThreadPoolExecutor(max_workers=len(targets)) as pool:
        futures = {
            name: pool.submit(_try_get, url, headers, timeout)
            for name, url in targets.items()
        }
        for name, fut in futures.items():
            try:
                res = fut.result(timeout=timeout + 1.0)
            except concurrent.futures.TimeoutError:
                res = None
            if res is not None:
                endpoints[name] = res

    return {
        "upstream_url": upstream_url,
        "provider": _classify(endpoints, upstream_url),
        "probed_at": int(time.time()),
        "endpoints": endpoints,
    }


def _classify(endpoints: dict, upstream_url: str) -> str:
    models_body = (endpoints.get("models") or {}).get("body") or {}
    model_ids = [m.get("id", "") for m in (models_body.get("data") or [])]

    # HF Router model ids carry a ``:<sub-provider>`` suffix.
    if any(":" in i for i in model_ids):
        return "hf-router"
    if endpoints.get("props") is not None:
        return "local-llama-server"

    info_body = (endpoints.get("info") or {}).get("body") or {}
    if isinstance(info_body, dict) and info_body.get("model_id"):
        return "tgi"

    version_body = (endpoints.get("version") or {}).get("body") or {}
    if isinstance(version_body, dict) and version_body.get("version"):
        return "vllm"

    if any(i.startswith(("gpt-", "o1-", "o3-", "o4-")) for i in model_ids):
        return "openai"

    return _hostname_fallback(upstream_url)


def refine_for_sub_provider(provider: str, model: str | None) -> str:
    """Surface HF Router's ``meta-llama/...:fireworks-ai`` pin as
    ``hf-router/fireworks-ai`` in the provider slug."""
    if provider == "hf-router" and model and ":" in model:
        return f"hf-router/{model.split(':', 1)[1]}"
    return provider

File: src/provider.rs
//! Identify the inference backend behind an upstream URL.
//!
//! Hostname classification ([`hostname_fallback`]) + the HF Router sub-provider pin
//! ([`refine_for_sub_provider`]) — what `run` and `export` use to slug a backend.
//! Live network introspection of the backend isn't implemented; the hostname slug
//! is enough for both paths.

use std::net::IpAddr;

/// Reverse proxies / custom domains won't match; the probe path catches those.
const HOSTNAME_TO_PROVIDER: &[(&str, &str)] = &[
    ("router.huggingface.co", "hf-router"),
    ("api.openai.com", "openai"),
    ("api.together.xyz", "together"),
    ("api.anthropic.com", "anthropic"),
    ("api.cerebras.ai", "cerebras"),
    ("api.fireworks.ai", "fireworks"),
    ("api.groq.com", "groq"),
];

/// Derive a provider slug from the upstream URL's hostname: known hosts map
/// directly; loopback / RFC1918 → `local`; otherwise the eTLD+1-ish second-level
/// label (`api.mycompany.com` → `mycompany`).
pub fn hostname_fallback(upstream_url: &str) -> String {
    let host = match url::Url::parse(upstream_url) {
        Ok(u) => u.host_str().unwrap_or("").to_lowercase(),
        Err(_) => String::new(),
    };
    // url keeps IPv6 hosts bracketed (`[::1]`); strip for the literal match below.
    let host = host.trim_start_matches('[').trim_end_matches(']').to_string();
    if host.is_empty() {
        return "unknown".to_string();
    }
    if let Some((_, p)) = HOSTNAME_TO_PROVIDER.iter().find(|(h, _)| *h == host) {
        return p.to_string();
    }
    if host == "localhost" || host == "::1" {
        return "local".to_string();
    }
    if let Ok(ip) = host.parse::<IpAddr>() {
        let is_local = ip.is_loopback() || matches!(ip, IpAddr::V4(v4) if v4.is_private());
        return if is_local { "local".to_string() } else { host };
    }
    let parts: Vec<&str> = host.split('.').collect();
    if parts.len() >= 2 {
        parts[parts.len() - 2].to_string()
    } else {
        host
    }
}

/// Surface HF Router's `meta-llama/...:fireworks-ai` pin as
/// `hf-router/fireworks-ai` in the provider slug.
pub fn refine_for_sub_provider(provider: &str, model: Option<&str>) -> String {
    if provider == "hf-router" {
        if let Some(m) = model {
            if let Some((_, sub)) = m.split_once(':') {
                return format!("hf-router/{sub}");
            }
        }
    }
    provider.to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn known_providers() {
        assert_eq!(hostname_fallback("https://router.huggingface.co/v1"), "hf-router");
        assert_eq!(hostname_fallback("https://api.openai.com/v1"), "openai");
        assert_eq!(hostname_fallback("https://api.together.xyz/v1"), "together");
        assert_eq!(hostname_fallback("https://api.fireworks.ai/v1"), "fireworks");
    }

    #[test]
    fn loopback_and_private() {
        assert_eq!(hostname_fallback("http://127.0.0.1:8000/v1"), "local");
        assert_eq!(hostname_fallback("http://localhost:8000/v1"), "local");
        assert_eq!(hostname_fallback("http://10.0.0.5:8000/v1"), "local");
        assert_eq!(hostname_fallback("http://192.168.1.42:8000/v1"), "local");
    }

    #[test]
    fn unknown_public() {
        assert_eq!(hostname_fallback("https://api.mycompany.com/v1"), "mycompany");
    }

    #[test]
    fn refine_pins_and_noops() {
        assert_eq!(
            refine_for_sub_provider("hf-router", Some("meta-llama/Llama-3.3-70B-Instruct:fireworks-ai")),
            "hf-router/fireworks-ai"
        );
        assert_eq!(
            refine_for_sub_provider("hf-router", Some("meta-llama/Llama-3.3-70B")),
            "hf-router"
        );
        assert_eq!(refine_for_sub_provider("local", Some("anything:fireworks-ai")), "local");
    }
}
